ambry-frontend/src/main/java/com/github/ambry/frontend/DeleteDatasetHandler.java

```diff
@@ -136,7 +143,20 @@ private void deleteDataset() throws RestServiceException {
         accountName = RestUtils.getHeader(restRequest.getArgs(), RestUtils.Headers.TARGET_ACCOUNT_NAME, true);
         containerName = RestUtils.getHeader(restRequest.getArgs(), RestUtils.Headers.TARGET_CONTAINER_NAME, true);
         datasetName = RestUtils.getHeader(restRequest.getArgs(), RestUtils.Headers.TARGET_DATASET_NAME, true);
-        //TODO: delete all the dataset version under the dataset.
+        List<DatasetVersionRecord>
+            datasetVersionRecordList = accountService.getAllValidVersion(accountName, containerName, datasetName);
+        for (DatasetVersionRecord record : datasetVersionRecordList) {
+          String version = record.getVersion();
+          RequestPath requestPath = getRequestPath(restRequest);
+          RequestPath newRequestPath =
+              new RequestPath(requestPath.getPrefix(), requestPath.getClusterName(), requestPath.getPathAfterPrefixes(),
+                  NAMED_BLOB_PREFIX + SLASH + accountName + SLASH + containerName + SLASH + datasetName + SLASH
+                      + version, requestPath.getSubResource(), requestPath.getBlobSegmentIdx());
+          // Replace RequestPath in the RestRequest and call DeleteBlobHandler.handle.
+          restRequest.setArg(InternalKeys.REQUEST_PATH, newRequestPath);
+          restRequest.setArg(Headers.DATASET_VERSION_QUERY_ENABLED, "true");
+          deleteBlobHandler.handle(restRequest, restResponseChannel, null);
+        }
         accountService.deleteDataset(accountName, containerName, datasetName);
```

Comment:
deleteBlobHandler.handle doesn't really guarantee the finish of the execution if you don't provide a callback and wait for the callback to be invoked.

And I found this to be werid, We can alway just use accountService to delete those dataset versions, why relay the operations to handler?

Reply:
I suppose we have to delete the dataset versions, delete the blob names and delete the blobs. Using deleteBlobHandler seems to be the only option here to reuse most of the code.

Reply:
Yeah, I'm updating the logic to use the recursive call back to delete the blob, then delete the dataset version then delete the dataset.